add cuda build

vicentebolea opened the pull request:
- **cuda: add cuda build**
- **MNT: Re-rendered with conda-smithy 3.53.3 and conda-forge-pinning 2025.12.05.12.42.42**

vicentebolea committed:
cuda: add cuda build

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint --conda-forge .` from the recipe directory. You can also examine the [workflow logs](https://github.com/conda-forge/conda-forge-webservices/actions/runs/19978349479) for more detail.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/19978349479](https://github.com/conda-forge/conda-forge-webservices/actions/runs/19978349479). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found it was in an excellent condition.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found some lint.

Here's what I've got...


For **recipe/recipe.yaml**:

 * ❌ This recipe is using a compiler, which now requires adding a build dependence on `${{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/19979039338](https://github.com/conda-forge/conda-forge-webservices/actions/runs/19979039338). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found it was in an excellent condition.

vicentebolea committed:
MNT: Re-rendered with conda-smithy 3.53.3 and conda-forge-pinning 2025.12.12.16.57.37

Other tools:
- conda-build 25.11.1
- rattler-build 0.54.0
- rattler-build-conda-compat 1.4.10

h-vetinari committed:
add cuda 12.9 migrator

h-vetinari committed:
remove misconfiguration for cuda_compiler_version

h-vetinari commented on recipe/recipe.yaml:
```diff
@@ -38,31 +39,51 @@ source:
+  number: 1
+  string: ${{ "cuda" ~ cuda_compiler_version ~ "_h" ~ hash ~ "_" ~ number if cuda_compiler_version != "None" else "cpu_h" ~ hash ~ "_" ~ number }}
+  skip:
+    - cuda_compiler_version != "None" and osx
```
this skip is redundant

h-vetinari commented on recipe/recipe.yaml:
```diff
@@ -38,31 +39,51 @@ source:
+  string: ${{ "cuda" ~ cuda_compiler_version ~ "_h" ~ hash ~ "_" ~ number if cuda_compiler_version != "None" else "cpu_h" ~ hash ~ "_" ~ number }}
+  skip:
+    - cuda_compiler_version != "None" and osx
+    - cuda_compiler_version != "None" and linux and aarch64
```
cuda works fine on aarch (including in cross-compilation) -- though no-one forces you to support more than you feel like

same for windows BTW

vicentebolea replied:
Viskores 1.1.0 does not currently support CUDA well in Windows

h-vetinari commented on recipe/recipe.yaml:
```diff
@@ -38,31 +39,51 @@ source:
+  variant:
+    use_keys:
+      - cuda_compiler_version
+    down_prioritize_variant: ${{ 1 if cuda_compiler_version != "None" else 0 }}
```
why do you want to de-prioritize the cuda variant? Better would be to down-prioritize the CPU variant and add a runtime-dependency on `__cuda` for the cuda build.

h-vetinari commented on recipe/recipe.yaml:
```diff
@@ -28,6 +28,7 @@ context:
     source
     worklet
   common_flags: -DCMAKE_BUILD_TYPE=Release -DViskores_ENABLE_HDF5_IO=ON -DViskores_USE_DEFAULT_TYPES_FOR_VTK=ON
+  cuda_flags: -DViskores_ENABLE_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=75
```
you could build for a broader set of architectures here; indeed for CUDA 12.6, our baseline set is usually:
`50;60;70;75;80;86;89;90`
(+ 10.0;12.0 for CUDA 12.9)

h-vetinari commented on recipe/recipe.yaml:
```diff
@@ -38,31 +39,51 @@ source:
+        - cmake --build build -j2
+        - cmake --install build
+        - Viskores_DIR=$PREFIX cmake -GNinja -S $SRC_DIR/examples/smoke_test/ -B smoke_test_build
+        - cmake --build smoke_test_build
```
IMO you should move this all to a proper `build.sh`, rather than duplicating two very similar build scripts.

vicentebolea replied:
I went another route and I factored out the diff of the two build instructions into a single one where we encode variables to support both of the behaviors

h-vetinari commented on recipe/recipe.yaml:
```diff
@@ -38,31 +39,51 @@ source:
     - ${{ pin_subpackage("viskores", upper_bound='x.x') }}
+  ignore_run_exports:
+    from_package:
+      - cuda-version
```
why?

h-vetinari replied:
Remove this
```suggestion
```

h-vetinari commented on recipe/recipe.yaml:
```diff
@@ -76,10 +97,15 @@ requirements:
       then: llvm-openmp
+  run:
+    - if: cuda_compiler_version != "None"
+      then: ${{ pin_compatible('cuda-version', upper_bound='x', lower_bound='x') }}
```
why?

vicentebolea replied:
https://github.com/conda-forge/cuda-feedstock/blob/main/recipe/doc/recipe_guide.md#cuda-enhanced-compatibility

h-vetinari replied:
This is wrong. Please remove.

h-vetinari replied:
```suggestion
```

vicentebolea committed:
address sugestions

vicentebolea commented:
@h-vetinari thanks for the review, when I try to rerender the changes in your last commit I get the following error message:

```
  zip_key entry c_stdlib_version in group frozenset({'fortran_compiler_version', 'cuda_compiler_version', 'c_stdlib_version', 'cxx_compiler_version', 'c_compiler_version'}) does not have any settings
```

command: `conda-smithy rerender -c auto`

h-vetinari commented:
> when I try to rerender the changes in your last commit I get the following error message:

My last commit is literally a successful rerender. I think what happened is that you fell into the short window where https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/8063 had been merged incorrectly (and broke the pinning). This has now been fixed and things should rerender fine again.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipe/recipe.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-recipe-manager`. The recipe can only be automatically migrated to the new v1 format if it is parseable by conda-recipe-manager.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/20309578559](https://github.com/conda-forge/conda-forge-webservices/actions/runs/20309578559). Examine the logs at this URL for more detail.</sub>

vicentebolea commented:
@h-vetinari thanks for the review, I have addressed all the points and I was able to rerender as you mentioned.

h-vetinari commented on recipe/recipe.yaml:
```diff
@@ -28,6 +28,13 @@ context:
+    -DViskores_ENABLE_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES='50;60;70;75;80;86;89;90'
+    {% else %}
+    -DViskores_ENABLE_CUDA=OFF
+    {% endif %}
```
I believe you're creating a multiline string including linebreaks here, which then doesn't work. Again, I suggest to move to a proper build script.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint --conda-forge .` from the recipe directory. You can also examine the [workflow logs](https://github.com/conda-forge/conda-forge-webservices/actions/runs/20310424655) for more detail.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/20310424655](https://github.com/conda-forge/conda-forge-webservices/actions/runs/20310424655). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipe/recipe.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-recipe-manager`. The recipe can only be automatically migrated to the new v1 format if it is parseable by conda-recipe-manager.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/20384872176](https://github.com/conda-forge/conda-forge-webservices/actions/runs/20384872176). Examine the logs at this URL for more detail.</sub>

vicentebolea committed:
MNT: Re-rendered with conda-smithy 3.54.1 and conda-forge-pinning 2025.12.19.22.26.27

Other tools:
- conda-build 25.11.1
- rattler-build 0.55.0
- rattler-build-conda-compat 1.4.10

vicentebolea commented:
@h-vetinari I had to go with 75,75-virtual for the cuda arch since it would OOM or timeout the build. Viskores is notorious for long build times and high memory usage. With this we get forward compatibilty thru PTX (JIT at runtime first run) for anything over 75 and already built cuda objects for 75.

h-vetinari commented on recipe/recipe.yaml:
```diff
@@ -28,6 +28,13 @@ context:
   common_flags: -DCMAKE_BUILD_TYPE=Release -DViskores_ENABLE_HDF5_IO=ON -DViskores_USE_DEFAULT_TYPES_FOR_VTK=ON
+  cuda_flags: >
+    {% if cuda_compiler_version != "None" %}
+    -DViskores_ENABLE_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES='75;75-virtual'
```
you could add a comment here about the architecture selection

vicentebolea replied:
@h-vetinari I will do it in a follow-up to avoid trigger long build times and further delay this

h-vetinari reviewed (approved):
LGTM, thanks!

vicentebolea commented:
@robertmaynard we now have a Viskores (formerly VTK-m) CUDA variant in conda-forge :tada:

vicentebolea commented:
@kmorel

kmorel commented:
Very cool.
